### PR TITLE
Add a stub Rust crate

### PR DESCRIPTION
**Pull request by erikjohnston**
This is a straw-man proposal for how we might want to integrate Rust into Synapse, and is not intended to land is is.

c.f. https://github.com/matrix-org/synapse/issues/12164

This bundles in Rust into Synapse proper, rather than e.g. having the rust code as a separate package. The advantage of this approach is that it allows very tight coupling between the Synapse and Rust code. The rust code is built and accessible at `synapse.synapse_rust.*`.

We achieve this by using a Poetry build script (which is a bit of an unstable feature https://github.com/python-poetry/poetry/issues/2740) to add the rust as an extension using [`setuptools-rust`](https://github.com/PyO3/setuptools-rust).

Using `poetry install` correctly builds the rust code.

Changes to the rust code have to be manually rebuilt by running `poetry install`.

Building of a source package via `poetry build -f sdist` builds a sdist that includes the rust source.

Building of wheels is done via the docker manylinux package and a script to use `poetry` to manually build wheels for each python version. Based on https://setuptools-rust.readthedocs.io/en/latest/building_wheels.html

### CODE REVIEW

**Comment by DMRobertson** [comment]
> We achieve this by using a Poetry build script (which is a bit of an unstable feature https://github.com/python-poetry/poetry/issues/2740)

Is there any particular version of poetry we need to use for this to make sense? We're split between 1.1.12 and 1.2.0bsomething in CI.

**Comment by erikjohnston** [comment]
> > We achieve this by using a Poetry build script (which is a bit of an unstable feature [python-poetry/poetry#2740](https://github.com/python-poetry/poetry/issues/2740))
> 
> Is there any particular version of poetry we need to use for this to make sense? We're split between 1.1.12 and 1.2.0bsomething in CI.

It seems to work fine with `1.1.13` which is what I have locally. Haven't tried the 1.2.0b yet

**Comment by erikjohnston** [comment]
This is now blocked on https://github.com/python-poetry/poetry/issues/6154 I think (or alternatively we add `setuptools_rust` as a normal dep). Frustratingly they fixed that issue for `poetry build` in 1.2.0b2 (which now just works), but `poetry install` doesn't work.

Otherwise, `pip install` works out of the box, so I think that is pretty much the only blocker?

**Comment by erikjohnston** [comment]
Slightly annoyingly poetry 1.1.14 doesn't let you specify a minimum version of setuptools, so unless you manually upgrade it `poetry install` doesn't work.

It also looks like they're not planning to fix https://github.com/python-poetry/poetry/issues/6154 imminently, so I've temporarily `setuptools_rust` to the dev-dependencies. Not sure if that is really the right place for it.

File: .github/workflows/release-artifacts.yml
Comment (reivilibre) [inline .github/workflows/release-artifacts.yml:94]: should whatever changed here be folded back into the backend-meta repo?

File: .gitignore
Comment (reivilibre) [inline .gitignore:66]: do we care about ... dylibs? dlls? Whatever those other OS people do :)

File: Cargo.toml
Comment (reivilibre) [inline Cargo.toml:5]: I wonder whether we should split things out into multiple crates later on, notably because the unit of compilation in Rust is a crate.
Definitely something we can afford to do at a later point in time, so no blocker, just inviting conversation.

File: pyproject.toml
Comment (reivilibre) [inline pyproject.toml:85]: do we need to include the `Cargo.toml` for the workspace root?
Sounds like maybe we don't *need* to but perhaps we should in case we wind up depending on it for options of some sort?

File: .github/workflows/release-artifacts.yml
Comment (DMRobertson) [inline .github/workflows/release-artifacts.yml:102]: I don't think you need `build` here, if we're using `poetry build` instead.

No objections to using `poetry build` instead of `python -m build` in the backend meta repo. (It should hopefully be equivalent anyway---I think build should invoke poetry-core via pyproject.toml?)

File: .github/workflows/release-artifacts.yml
Comment (erikjohnston) [reply .github/workflows/release-artifacts.yml:94]: Ah yup, probably. The change is using `poetry build` rather than `pip build`

File: .gitignore
Comment (erikjohnston) [reply .gitignore:66]: Yeah, probably I guess? Though do we support other OSes?

File: Cargo.toml
Comment (erikjohnston) [reply Cargo.toml:5]: Yeah, possibly. Though for dev builds incremental compilation is now a thing.

File: pyproject.toml
Comment (erikjohnston) [reply pyproject.toml:85]: oooh, yes

File: .github/workflows/release-artifacts.yml
Comment (erikjohnston) [reply .github/workflows/release-artifacts.yml:102]: For whatever reason `python -m build` wasn't installing the build deps iirc?

File: tests/test_rust.py
Comment (reivilibre) [inline tests/test_rust.py:1]: More of a style question / point to think about: do we want to write unit tests in Rust (and then use `cargo test` to run them in CI?)

Trial tests still seem to make sense for testing the external behaviour of the Rust module.

**Review by reivilibre — DISMISSED** [review]
Looks promising to me; I would double check that the sdist is good enough to use as an installation source before merging.

Will re-add to queue since you asked for general opinions on this and mine is but one.

**Comment by erikjohnston** [comment]
We also need to figure out a minimum supported rustc version

File: .gitignore
Comment (clokep) [reply .gitignore:66]: Yes, I'm on macOS (and others who occasionally contribute to Synapse use macOS too) -- are you mostly just hoping I run this and make sure it doesn't break / what other files we might need to ignore?

Comment (erikjohnston) [reply .gitignore:66]: Yeah, pretty much

**Comment by erikjohnston** [comment]
The complement tests are failing as its using the released docker image to build Synapse, which doesn't yet have rust installed.

File: .gitignore
Comment (clokep) [reply .gitignore:66]: On macOS this did seem to only generate a `.so` file, but it was under the `build/` directory. This was when installing an editable install via `pip`, so might be OK if you use poetry?

Comment (erikjohnston) [reply .gitignore:66]: `/build/` is ignored separately fwiw, so thats fine.

File: rust/src/lib.rs
Comment (clokep) [inline rust/src/lib.rs:1]: (Somewhat leaving this comment in a random spot...)

Will integrating Rust code this way end up with the same issues we ran into on #10857 -- which I think boils down to "what do we do about metrics in Rust code?"

Comment (erikjohnston) [reply rust/src/lib.rs:1]: Yes, though we have that problem today with other native modules. The scale of the problem however is different: the state compressor happens in the background and consumed multiple cores, while most of the native code we run is "inline" on the main thread and doesn't consume much CPU.

We probably will want to figure something out here if we end up doing lots of CPU in rust, but I don't see it as blocking.

Comment (clokep) [reply rust/src/lib.rs:1]: That makes sense, thanks for the explanation!

**Comment by erikjohnston** [comment]
Next steps:

- [ ] Update install instructions for source checkouts
- [ ] Add CI for uploading all the appropriate wheels. We should be able to steal from how [cryptography do it](https://github.com/pyca/cryptography/blob/main/.github/workflows/wheel-builder.yml).
- [ ] Add generic Rust CI for `cargo test|fmt|check|clippy` etc

File: docker/Dockerfile
Comment (DMRobertson) [inline docker/Dockerfile:56]: Did this end up happening in poetry 1.2?

Comment (erikjohnston) [reply docker/Dockerfile:56]: Yes, except 1.2 pulls it in automatically (there is a fun circular dependency between poetry and poetry export plugin...)

**Comment by erikjohnston** [comment]
I think this is now ready for review! It's currently based on https://github.com/matrix-org/synapse/pull/13707

File: .github/workflows/release-artifacts.yml
Comment (erikjohnston) [inline .github/workflows/release-artifacts.yml:112]: We should also limit the number of wheels we build for PRs.

File: pyproject.toml
Comment (babolivier) [inline pyproject.toml:56]: I assume this is referring to the name of the checkout of the repo rather than the repo's `synapse` directory? If so can you add a comment about this because I would find it a bit confusing.

File: docs/development/contributing_guide.md
Comment (babolivier) [inline docs/development/contributing_guide.md:121]: This reads weird, either the "either" or the parentheses should go imo.

File: docs/deprecation_policy.md
Comment (babolivier) [inline docs/deprecation_policy.md:44]: > In general, Synapse will unlikely ever require the absolute latest Rust version

This feels like it contradicts what is said above:

> For any given release the minimum required version may be bumped up to the latest released Rust version

File: docker/complement/conf/start_for_complement.sh
Comment (babolivier) [inline docker/complement/conf/start_for_complement.sh:107]: Can you add a comment explaining why we're doing this please?

File: docs/development/contributing_guide.md
Comment (reivilibre) [inline docs/development/contributing_guide.md:123]: Maybe worth thinking down the line: is there a way that someone who only wants to touch the Python code can short-cut this setup step / requirement of a rust compiler, e.g. by first installing a wheel?
Certainly not a blocker, just a though

File: pyproject.toml
Comment (reivilibre) [inline pyproject.toml:142]: do we need to warn packagers?

File: .github/workflows/release-artifacts.yml
Comment (babolivier) [inline .github/workflows/release-artifacts.yml:92]:
```suggestion
  build-wheels:
```

For consistency

**Review by babolivier — CHANGES_REQUESTED** [review]

File: rust/Cargo.toml
Comment (reivilibre) [inline rust/Cargo.toml:7]:
```suggestion
rust-version = "1.61.0"
```

I think we said 1.61.0 for now in #synapse-dev.

it even bit me :-)
```
error: package `synapse v0.1.0 (/tmp/xxx/synapse/rust)` cannot be built because it requires rustc 1.63.0 or newer, while the currently active rustc version is 1.61.0
```

**Review by reivilibre — APPROVED** [review]
This works for me and I believe it to be reasonable. LGTM

File: pyproject.toml
Comment (erikjohnston) [reply pyproject.toml:142]: Woops, that is now spurious.

File: docs/development/contributing_guide.md
Comment (erikjohnston) [reply docs/development/contributing_guide.md:123]: There's no good support for doing that TBH. I think we'd have to have the rust as a separate python package to make that work.

File: pyproject.toml
Comment (erikjohnston) [reply pyproject.toml:142]: woops, that isn't needed anymore

File: pyproject.toml
Comment (erikjohnston) [reply pyproject.toml:56]: Oooh, turns out that isn't needed anymore. :tada:

File: docs/deprecation_policy.md
Comment (erikjohnston) [reply docs/deprecation_policy.md:44]: It's a hard one, as we want to reserve the right to do so, but will likely never actually do it except in exceptional circumstances? Not sure how better to frame it.

File: docker/complement/conf/start_for_complement.sh
Comment (erikjohnston) [reply docker/complement/conf/start_for_complement.sh]: I think that is unnecessary now, will remove and see.

File: docs/deprecation_policy.md
Comment (babolivier) [reply docs/deprecation_policy.md:44]: I would say either remove the bit about Synapse's requirement from this sentence (ie turn it into something like "Because of this, introducing a formal policy is hard given the constraints of the ecosystem."), or reword the first half into something like "In practice, Synapse should only rarely require the absolute latest version of Rust". My issue with the current state of this sentence is that I read it as that in theory it's possible that requiring the latest version of Rust happens however in practice users should expect it to never happen, which isn't really the message we want to communicate.

File: .github/workflows/release-artifacts.yml
Comment (babolivier) [inline .github/workflows/release-artifacts.yml:103]: Do we have anything to alert us if the macos wheels fail to build? I feel like simply having the green tick turn into a red cross on https://github.com/matrix-org/synapse/tree/develop is easy to miss, so I would say ideally we would need something like we have for the twisted trunk job.

**Comment by erikjohnston** [comment]
CI should hopefully be fixed by https://github.com/matrix-org/synapse/pull/13725

File: docs/deprecation_policy.md
Comment (erikjohnston) [reply docs/deprecation_policy.md]: I've updated the wording a bit to make it a bit more balanced, maybe?

File: .github/workflows/release-artifacts.yml
Comment (erikjohnston) [reply .github/workflows/release-artifacts.yml:103]: Mmm, maybe we should just have a generic issue if the develop / release tests fail? Rather than special casing each one.

Note we already only tests small subsets of debian version when building debs on PRs, so this is in line with that.

File: docs/deprecation_policy.md
Comment (babolivier) [reply docs/deprecation_policy.md]: Much better yep!

File: debian/rules
Comment (babolivier) [inline debian/rules:54]: Not sure about this one - I'm not familiar enough with debian packages to answer confidently. But since the CI job is passing it should be fine?

File: .github/workflows/tests.yml
Comment (babolivier) [inline .github/workflows/tests.yml:137]:
```suggestion
    needs: linting-done
```

(just making sure we don't forget before hitting merge)

File: docs/deprecation_policy.md
Comment (babolivier) [inline docs/deprecation_policy.md:44]:
```suggestion
Synapse team will try to avoid updating the dependency on Rust to the absolute latest version, but
```

**Review by babolivier — APPROVED** [review]
Otherwise I'm happy if CI is happy :slightly_smiling_face:

File: debian/rules
Comment (babolivier) [reply debian/rules:54]: > But since the CI job is passing it should be fine?

Oh actually it doesn't look like we're testing building the debs in CI? (or the job has been renamed and I can't find it on this PR's status)

File: .github/workflows/release-artifacts.yml
Comment (DMRobertson) [inline .github/workflows/release-artifacts.yml:103]: Do we even need wheels for mac? Is the point to make it easier for contributors using macs to make Python-only changes?

File: .github/workflows/release-artifacts.yml
Comment (DMRobertson) [inline .github/workflows/release-artifacts.yml:138]: Is this just here to avoid spawning a new job?

Comment (DMRobertson) [reply .github/workflows/release-artifacts.yml:138]: Actually, aren't we going to be building the same sdist once for every platform here?

File: .github/workflows/tests.yml
Comment (DMRobertson) [inline .github/workflows/tests.yml:146]: Is it worth pinning to a specific version of rustc here, to help keep builds more reproducible?

File: .github/workflows/tests.yml
Comment (DMRobertson) [inline .github/workflows/tests.yml:184]: Intentional?

File: .github/workflows/tests.yml
Comment (DMRobertson) [inline .github/workflows/tests.yml:371]: Wrong `name` here?

File: build_rust.py
Comment (DMRobertson) [inline build_rust.py:8]: Is mypy going to check this file? Type hints would be nice, but I won't block this on having them.

File: docker/Dockerfile-dhvirtualenv
Comment (DMRobertson) [inline docker/Dockerfile-dhvirtualenv:89]:
```suggestion
# Install rust and ensure it's in the PATH
```

File: pyproject.toml
Comment (DMRobertson) [inline pyproject.toml:325]: I guess build reads pyproject.toml and invokes `poetry-core` to build the wheel?

File: rust/Cargo.toml
Comment (DMRobertson) [inline rust/Cargo.toml:5]: Should this match the version specified in pyproject.toml?

File: rust/Cargo.toml
Comment (DMRobertson) [inline rust/Cargo.toml:7]: Is this the version you decided on? I thought I saw talk of 1.58 in synase-dev but might be confused

File: stubs/synapse/synapse_rust.pyi
Comment (DMRobertson) [inline stubs/synapse/synapse_rust.pyi:1]: Is there a way to automatically generate this (in the future)?

Also: Python's `int`s are bigger than Rust's `usize`s. What happens if we call `sum_as_string(2**100, 2**200)` with bigger-than-usize ints?

Comment (DMRobertson) [reply stubs/synapse/synapse_rust.pyi:1]: Also: does `stubs/synapse/__init__.pyi` clobber the annotations in `synapse/__init__.py`? Oh, that file is empty except for a copyright notice.

File: tests/test_rust.py
Comment (DMRobertson) [inline tests/test_rust.py:9]: Can we add `test_rust` to the list of files typechecked by mypy please?

File: debian/rules
Comment (erikjohnston) [reply debian/rules:54]: Oh gah, I've broken the `github/workflows/release-artifacts.yml` workflow :facepalm:

File: .github/workflows/release-artifacts.yml
Comment (erikjohnston) [reply .github/workflows/release-artifacts.yml:103]: TBH, I mostly just included mac wheels because cryptography supports them. I don't think its easy to use an existing wheel to bootstrap a source checkout, alas.

File: .github/workflows/release-artifacts.yml
Comment (erikjohnston) [reply .github/workflows/release-artifacts.yml:138]: Oh, true. I should put it in a separate job.

File: .github/workflows/tests.yml
Comment (erikjohnston) [reply .github/workflows/tests.yml:184]: Debugging hangover :facepalm:

File: .github/workflows/tests.yml
Comment (erikjohnston) [reply .github/workflows/tests.yml:371]: Where has nightly come from, but yes

File: pyproject.toml
Comment (erikjohnston) [reply pyproject.toml:325]: Yes it does.

File: rust/Cargo.toml
Comment (erikjohnston) [reply rust/Cargo.toml:5]: I don't believe it has any affect, its just a mandatory field in Rust. TBH I would rather not try to keep the two in sync.

File: rust/Cargo.toml
Comment (erikjohnston) [reply rust/Cargo.toml:7]: Yeah, I think 1.61.0 is a reasonable one. It's from May (so four months old) and is what the latest firefox requires.

Its mostly pulled out of thin air though.

File: rust/Cargo.toml
Comment (DMRobertson) [reply rust/Cargo.toml:5]: I suppose we're not planning on publishing this to crates.io or any other registry.

Maybe drop a quick comment e.g.

```toml
# dummy version. See pyproject.toml for the Synapse's version number.
```

> TBH I would rather not try to keep the two in sync.

Agreed.

File: stubs/synapse/synapse_rust.pyi
Comment (erikjohnston) [reply stubs/synapse/synapse_rust.pyi:1]: > Is there a way to automatically generate this (in the future)?

I don't believe anyone has written any tooling for it, annoyingly.

> Also: Python's ints are bigger than Rust's usizes. What happens if we call sum_as_string(2**100, 2**200) with bigger-than-usize ints?

It will raise an `OverflowError`. In general `pyo3` will raise an exception whenever it tries to convert from a python to rust type.

> Also: does stubs/synapse/__init__.pyi clobber the annotations in synapse/__init__.py? Oh, that file is empty except for a copyright notice.

Empirically `mypy` does the right thing here and uses `synapse/__init__.py` (I just tested it by doing `reveal_type(synapse.__version__) `

File: .github/workflows/release-artifacts.yml
Comment (DMRobertson) [reply .github/workflows/release-artifacts.yml:103]: We write [here](https://matrix-org.github.io/synapse/latest/setup/installation.html?highlight=windows#windows) that

> Running Synapse natively on Windows is not officially supported.

No such wording for [mac](https://matrix-org.github.io/synapse/latest/setup/installation.html?highlight=mac#macos) though.

I guess leave it in since it's only for release builds. :shrug:

File: debian/rules
Comment (babolivier) [reply debian/rules:54]: Ah that'd explain it. It's annoying GitHub doesn't show a warning or error on the PR when that's the case.

File: build_rust.py
Comment (DMRobertson) [reply build_rust.py:8]: Alternatively stick it in `scripts-dev` and it'll be covered with everything else :angel:

File: debian/rules
Comment (erikjohnston) [reply debian/rules:54]: Empirically the debs seem to work when I tested them, and the build script does run the unit tests, so I think we should be good.

File: build_rust.py
Comment (erikjohnston) [reply build_rust.py:8]: I don't think it can go in scripts-dev as it needs to be in the source to build the package?

File: tests/test_rust.py
Comment (erikjohnston) [reply tests/test_rust.py:9]: All tests are typechecked by default I think?

Comment (DMRobertson) [reply tests/test_rust.py:9]: Oh sorry, I was thinking of `disallow_untyped_defs` which isn't on by default.

But let's drop that for now.

File: .github/workflows/release-artifacts.yml
Comment (DMRobertson) [inline .github/workflows/release-artifacts.yml:139]: Does this work? Do you need to setup-python or `pip install build`?

Comment (erikjohnston) [reply .github/workflows/release-artifacts.yml:139]: Sigh, good point. Lets test that.

Comment (DMRobertson) [reply .github/workflows/release-artifacts.yml:139]: Erm, I think you're missing an `actions/checkout ` too.

**Review by DMRobertson — APPROVED** [review]
Other than building the sdist lgtm. (I have no idea about the debian packaging.)

Something to think about for the future anticipate us writing unit tests in rust at all?

**Comment by erikjohnston** [comment]
> Something to think about for the future anticipate us writing unit tests in rust at all?

Yeah, I plan to do a follow up PR that adds Rust specific CI (i.e. `cargo fmt|check|clippy|test`). Just don't want to clutter this PR up even more

File: .github/workflows/release-artifacts.yml
Comment (erikjohnston) [reply .github/workflows/release-artifacts.yml:146]: Yup, fixed now!

File: .github/workflows/release-artifacts.yml
Comment (DMRobertson) [inline .github/workflows/release-artifacts.yml:135]: Hmm. Should we be building the sdist on PRs? (I'm not sure it picks anything hugely important up?)

File: .github/workflows/tests.yml
Comment (jplatte) [inline .github/workflows/tests.yml:146]: After toolchain installation, I recommend adding <code>- uses: [Swatinem/rust-cache@v2](https://github.com/Swatinem/rust-cache#readme)</code> as a step, which will take care of caching dependencies.

File: Cargo.toml
Comment (jplatte) [inline Cargo.toml:1]: No corresponding `Cargo.lock`?

File: docs/deprecation_policy.md
Comment (jplatte) [inline docs/deprecation_policy.md:41]: This should be "the Rust project", the foundation only does trademark, financing and PR stuff, it's not involved with maintenance / suppport or such.

**Review by jplatte — COMMENTED** [review]
A minute too late 😅

File: Cargo.toml
Comment (erikjohnston) [reply Cargo.toml:1]: Oh, good point!

**Comment by NickCao** [comment]
I noticed that there is no `Cargo.lock` in the release tarball or 1.68rc1, is this intended or will be addressed in the final release?

**Comment by DMRobertson** [comment]
> I noticed that there is no `Cargo.lock` in the release tarball or 1.68rc1, is this intended or will be addressed in the final release?

It was an oversight, see #13858, which will be included in the final 1.68 release.

**Comment by ri0t** [comment]
Upgrading a roughly year old installation on Debian (from git) yields:
```
# synctl restart
synapse.app.homeserver not running
Traceback (most recent call last):
  File "/usr/lib/python3.9/runpy.py", line 188, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "/usr/lib/python3.9/runpy.py", line 111, in _get_module_details
    __import__(pkg_name)
  File "/home/synapse/synapse/synapse/__init__.py", line 23, in <module>
    from synapse.util.rust import check_rust_lib_up_to_date
  File "/home/synapse/synapse/synapse/util/rust.py", line 20, in <module>
    from synapse.synapse_rust import get_rust_file_digest
ModuleNotFoundError: No module named 'synapse.synapse_rust'
```

I have installed rustc via apt-get (not soo cool to have to, version 1.60 - maybe too new?) then did `pip install --upgrade .` which ran without errors/problems. Am now installing tox et al to maybe get a few insights.

**Comment by ri0t** [comment]
Nope, not even doing a clean (pypi based) install in a new venv doesn't work, yields the same problem.

**Comment by ri0t** [comment]
Ah, well. A very thorough cleanup (deleting everything..!) with partial restoring from backup cleared it up.

**Comment by DMRobertson** [comment]
> ModuleNotFoundError: No module named 'synapse.synapse_rust'

This error appears in at least two circumstances:

1. You have not build the Rust part of Synapse.
2. You have built the Rust part of Synapse, but you are running Synapse directly from its source tree. This won't work because Python is importing Synapse from the source checkout, and not where it's actually installed.

I suspect the latter in your case.

**Comment by babolivier** [comment]
Also: please use [`#synapse:matrix.org`](https://matrix.to/#/#synapse:matrix.org) for support rather than trying to revive long-closed PRs.

**Pull request merged**
